File: update_drawers.py
import re

path = r'C:\website\flutter-project\lib\screens\dashboard_screen.dart'
with open(path, 'r', encoding='utf-8') as f:
    content = f.read()

# 1. Add fetch flags to State class
if 'bool _openPlaysFetched = false;' not in content:
    content = content.replace('bool _isLoading = true;', 'bool _isLoading = true;\n  bool _openPlaysFetched = false;\n  bool _openChallengesFetched = false;\n  bool _pasaloCourtsFetched = false;')

# 2. Remove fetches from _loadUserData
content = content.replace('        _fetchOpenPlays();\n', '')
content = content.replace('        _fetchOpenChallenges();\n', '')
content = content.replace('        _fetchPasaloCourts();\n', '')

def replace_section(content, method_name, title, fetch_method, flag_var, list_var):
    # Find start of method
    start_idx = content.find(f'Widget {method_name}() {{')
    if start_idx == -1: return content
    
    # Find the end of method by brace matching
    brace_count = 0
    in_method = False
    end_idx = -1
    for i in range(start_idx, len(content)):
        if content[i] == '{':
            brace_count += 1
            in_method = True
        elif content[i] == '}':
            brace_count -= 1
            if in_method and brace_count == 0:
                end_idx = i
                break
                
    method_content = content[start_idx:end_idx+1]
    
    # We want to replace the `return Column(`... up to `children: [`
    # with the ExpansionTile, and then at the very end of method replace `);` with `);` and close the ExpansionTile
    
    # First, find the 'return Column('
    col_start = method_content.find('return Column(')
    if col_start == -1: return content
    
    # Let's replace the whole top part of the Column
    # (from 'return Column' down to the first 'if')
    if_start = method_content.find(f'if ({list_var}.isEmpty)')
    if if_start == -1: return content
    
    # We will inject the ExpansionTile header and the loading state
    # Notice we use Theme.of(context).copyWith(dividerColor: Colors.transparent) to hide the border
    new_header = f"""return Theme(
      data: Theme.of(context).copyWith(dividerColor: Colors.transparent),
      child: ExpansionTile(
        initiallyExpanded: false,
        onExpansionChanged: (expanded) {{
          if (expanded && !{flag_var}) {{
            {fetch_method}();
            setState(() {{ {flag_var} = true; }});
          }}
        }},
        title: Text(
          '{title}',
          style: TextStyle(fontFamily: 'Poppins', fontSize: 22, fontWeight: FontWeight.w600, letterSpacing: -0.5, color: AppColors.deepTeal),
        ),
        children: [
          if (!{flag_var})
            Padding(padding: EdgeInsets.all(20), child: CircularProgressIndicator(color: AppColors.accentLime))
          else """
    
    # Replace from 'return Column(' to the 'if (_something.isEmpty)'
    top_replaced = method_content[:col_start] + new_header + method_content[if_start:]
    
    # Now fix the end of the method
    # The original method ended with:
    #       ],
    #     );
    #   }
    # We need to change it to:
    #       ],
    #     ),
    #   );
    # }
    bottom_replaced = top_replaced.rsplit(';', 1)
    if len(bottom_replaced) == 2:
        top_replaced = bottom_replaced[0] + '),\n    );' + bottom_replaced[1]
        
    return content[:start_idx] + top_replaced + content[end_idx+1:]

# Apply replacements
content = replace_section(content, '_buildOpenPlaysSection', 'Active Open Plays', '_fetchOpenPlays', '_openPlaysFetched', '_openPlays')
content = replace_section(content, '_buildOpenChallengesSection', 'Active Open Challenges', '_fetchOpenChallenges', '_openChallengesFetched', '_openChallenges')
content = replace_section(content, '_buildPasaloCourtsSection', 'Pasalo Courts', '_fetchPasaloCourts', '_pasaloCourtsFetched', '_pasaloCourts')

with open(path, 'w', encoding='utf-8') as f:
    f.write(content)
print("Dashboard completely updated with ExpansionTiles!")

File: update_drawers.js
const fs = require('fs');

const path = 'C:/website/flutter-project/lib/screens/dashboard_screen.dart';
let code = fs.readFileSync(path, 'utf8');

// 1. Remove fetches from _loadUserData
code = code.replace('_fetchOpenPlays();\n', '');
code = code.replace('_fetchOpenChallenges();\n', '');
code = code.replace('_fetchPasaloCourts();\n', '');

// 2. Add fetched flags to DashboardScreenState
if (!code.includes('bool _openPlaysFetched')) {
    code = code.replace('bool _isLoading = true;', 'bool _isLoading = true;\n  bool _openPlaysFetched = false;\n  bool _openChallengesFetched = false;\n  bool _pasaloCourtsFetched = false;');
}

// Helper to replace sections with ExpansionTile
function replaceSection(code, sectionName, title, fetchMethod, fetchedFlag, listVar, emptyText) {
    const startStr = `  Widget _build${sectionName}Section() {\n    return Column(`;
    
    // Find the end of the method
    let startIndex = code.indexOf(startStr);
    if (startIndex === -1) {
        console.log(`Could not find ${sectionName}`);
        return code;
    }
    
    // Simplistic bracket matching to find end of method
    let openBrackets = 0;
    let i = startIndex + `  Widget _build${sectionName}Section() {`.length;
    let endIndex = -1;
    for (; i < code.length; i++) {
        if (code[i] === '{') openBrackets++;
        if (code[i] === '}') {
            if (openBrackets === 0) {
                endIndex = i + 1;
                break;
            }
            openBrackets--;
        }
    }
    
    const originalMethod = code.substring(startIndex, endIndex);
    
    // We need to extract the empty state and the list view from the original
    const emptyStateStart = originalMethod.indexOf(`if (${listVar}.isEmpty)`);
    const elseStart = originalMethod.indexOf(`else`, emptyStateStart);
    const listStart = originalMethod.indexOf(`SizedBox(`, elseStart) !== -1 ? originalMethod.indexOf(`SizedBox(`, elseStart) : originalMethod.indexOf(`ListView`, elseStart);
    
    // Because parsing the inner children of the column accurately via string manipulation is hard,
    // let's just create a completely fresh method for the ExpansionTile, and we just inject the list building logic.
    // Wait, the list building logic (the cards) are different for each.
    // Let's just find `if (${listVar}.isEmpty)` and keep everything from there to the end of the children array.
    
    // Better idea: replace the start of the Column with the ExpansionTile
    let newMethod = originalMethod.replace(
        /return Column\([\s\S]*?children: \[[\s\S]*?Padding\([\s\S]*?child: Row\([\s\S]*?Active Open[\s\S]*?SizedBox\(height: 16\),/i,
        `return Theme(
      data: Theme.of(context).copyWith(dividerColor: Colors.transparent),
      child: ExpansionTile(
        initiallyExpanded: false,
        onExpansionChanged: (expanded) {
          if (expanded && !${fetchedFlag}) {
            ${fetchMethod}();
            setState(() { ${fetchedFlag} = true; });
          }
        },
        title: Text(
          '${title}',
          style: TextStyle(fontFamily: 'Poppins', fontSize: 22, fontWeight: FontWeight.w600, letterSpacing: -0.5, color: AppColors.deepTeal),
        ),
        children: [
          if (!${fetchedFlag}) 
            Padding(padding: EdgeInsets.all(20), child: CircularProgressIndicator(color: AppColors.accentLime))
          else if (${listVar}.isEmpty)
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 24, vertical: 8),
              child: Container(
                padding: EdgeInsets.all(16),
                decoration: BoxDecoration(
                  color: Colors.grey.shade50,
                  borderRadius: BorderRadius.circular(12),
                  border: Border.all(color: Colors.grey.shade200),
                ),
                child: Row(
                  children: [
                    CustomPaddleIcon(color: Colors.grey.shade400),
                    SizedBox(width: 16),
                    Expanded(
                      child: Text(
                        '${emptyText}',
                        style: TextStyle(color: Colors.grey.shade600, fontSize: 13),
                      ),
                    ),
                  ],
                ),
              ),
            )
          else`
    );
    
    // Now we need to remove the original empty state check and just keep the `else` block content.
    // Wait, regex might destroy the card. 
    return code; // aborting this approach for a safer one
}

// Safer approach: I will just use manual replace on the exact text of the sections.
fs.writeFileSync('update_script_temp.js', '/* will run custom logic */');
